File: js/packages/teams-ai/src/TestTurnState.ts
import { TurnState } from './TurnState';
import { MemoryStorage, TurnContext } from 'botbuilder';

/**
 * A test version of the TurnState class used by unit tests.
 */
export class TestTurnState extends TurnState {
    /**
     * @private
     */
    private constructor() {
        super();
    }

    /**
     * Creates a new `TestTurnState` instance.
     * @param context Context for the current turn of conversation with the user.
     * @param testState Optional. State to initialize the new instance with.
     * @param testState.user
     * @param testState.conversation
     * @param testState.temp
     * @returns Created `TestTurnState` instance.
     */
    public static async create(
        context: TurnContext,
        testState?: { user?: Record<string, any>; conversation?: Record<string, any>; temp?: Record<string, any> }
    ): Promise<TestTurnState> {
        const storage = new MemoryStorage();
        const state = new TestTurnState();
        await state.load(context, storage);
        if (testState) {
            if (testState.user) {
                state.user = testState.user;
            }
            if (testState.conversation) {
                state.conversation = testState.conversation;
            }
            if (testState.temp) {
                state.temp = testState.temp as any;
            }
        }
        return state;
    }
}

File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/State/TestTurnState.cs
using Microsoft.Bot.Builder;

namespace Microsoft.Teams.AI.State
{
    /// <summary>
    /// Turn state used for testing purposes.
    /// </summary>
    public class TestTurnState : TurnState
    {
        private TestTurnState() : base() { }

        /// <summary>
        /// Returns a test turn state based on loaded configured options.
        /// </summary>
        /// <param name="turnContext">Current turn context.</param>
        /// <param name="testState">Options for TestTurnState class.</param>
        /// <returns></returns>
        public static async Task<TestTurnState> Create(ITurnContext turnContext, TestTurnStateOptions testState)
        {
            MemoryStorage storage = new();
            TestTurnState state = new();
            await state.LoadStateAsync(storage, turnContext);

            if (testState != null)
            {
                if (testState.User != null)
                {
                    state.User = testState.User;
                }
                if (testState.Conversation != null)
                {
                    state.Conversation = testState.Conversation;
                }
                if (testState.Temp != null)
                {
                    state.Temp = testState.Temp;
                }

            }
            return state;
        }
    }
}
